### ISSUE TITLE
infinite loop during key loading

### ISSUE THREAD

**Issue by richardmss** [SecurityInnovation/PGPy#252]
I'm loading a pgp key, however the library seems to be in an infinite loop. Here is how to reproduce:
```
PGPKey.from_blob(b'\x99\x01\r\x04P\xaa+h\x01\x08\x00\xa2\xd3\xb4\xb5\n\xab\xd5Z\xed\xd6\xf3\x1eQ^\xb3\xddJ\xd4\xbbD\xb1\x90\xa4\xe1q\xbb@\xf2=\xb9g\xb0B\x8e/\xc4!\xa97\xd28\xa6\xd9\xdc\xd5|\xb1\xe5\x9c\x08\x98\xd5j\xa9\xb1.,4~?B\x8b}\x83Z\xf3R\xa0\x96M\x99 \xd6\x9a\x95|\x81\xc8J\x90:\xb2\xa0\xd1\xca\x0fl\xf0\xd1\xe0R\x933\xa7\x83\xe1\n\xd4y\x12\xf3\xd0A@\xb2\xb2\x05E2\xeaA^\x8bt\xfe7\xc0\xd7a\xc8\xc4\x1c\xd8\xab\xce\xee\x16D\x85\x8f,}\x85d}!]zM\xbc\xab\xefH\x18\x85aQ\xa5Q\xb655"w\xcb\xb3\xde\x7f\xecm\xe9\xc0\x95\x12V\x7fc1n\xc9V\xac\xe5\x97\x8d\x9a\xd8\xa10"\xa1?_\xd8\x98\xb4N\x99\xb8\xcdUkq\x95\xdby\xf7\xe6n\x8eV\xe9-5{\xc7\xe1a\x1d\xf2e\xcb_|Q\xf6\x1a\xd7\x96\x95\xdb\xf1\xea\xea\xccl\x9d\xac\x84/\xc4\x1ez\xb4t\x9a\x8c\xaa\x08;U\xd3\xddD;\xea\x9d}\x02\x02\rsM\t\x0b\xb7\x00\x11\x01\x00\x01\xb41Standardsoftware <standardsoftware@volkswagen.de>\x89\x01%\x04\x13\x01\x02\x00\x19\x05\x02P\xaa+h\t\x10\x00\xf7\xf3aM\x98xv\x02\x00\t\x05\t\x03\xc2g\x00\x00\x00\r\xb7\x07\xff]\xb2\xb6\xca\x1e\xdc\xbf\x1e\x83\xd0Y\xbbgs\xcf\x92\x1dP\x98!\xe4w\xc7\rrK:Ef\x04\xc0\x96V~\xc1\x83\xa3o\x0b\x98\x95<l2RZ\xaa\xc0\r\xa7\xa6\xdfTs\xa9\rR-\x05!\x0c\xc4\xd4\xcf\xb0B\xb7\xaeR\xae\xc6\x0cM\xf8\x80\xf2\x80\xf6%Wu\xf2\x1c-\xa3\nO\xc3\x9ba\x0e \xf3~,\xf6.d\xb2\xe0v\xb8\tR%\xf4\x01U\xffi\x8a\xad\x90\xd7\xb0O\xd5%[V\xbb6\xa1\x13\xa0{\xf0\xd2\x14X\xa61\xdd\xf0 \xb2j\xcbko\xaf\x9b\x7fp3\x17\x7f\x85\xbd\xc1Y7\xf8e\x08\x84\x81V\xd2T\xa7\x91\xeaw\x8f\x1f\xc5\x9cX\xf3\xcd\x96\x8d\xa6p2fR\xf7[|\x12\xff\xab>\xe2\xd4\x12j\xf9[y\xcc\xc3\xc0\xe4L\xf3\xc0\x00\x95z\x98\xcdx\xb6\x14\xcar#T\xc7\xb3 x~\x8dD\xcby\xd5\xa7\xe0[\xb7 \x1fi{\xcfo\x86i\xbc\x81\x0b\xaf\x0eE\x89p\xcd\xdd\xb4\xf8\xad\xf2\x02Yb\xf0\x02\xb9\xd2\x92\x02')
```

**Comment by dkg** [comment]
i can replicate this with the current git master, unfortunately.

The certificate being parsed is 
[volks.txt](https://github.com/SecurityInnovation/PGPy/files/3458559/volks.txt).  The same non-terminating call appears to happen with the ascii-armored form as well.

**Comment by dkg** [comment]
I think it has to do with the certificate's selfsig's subpacket of type 0 (reserved, according to RFC 4880) with length1, but i confess i'm a bit lost in the twisty maze of dispatch and abstract baseclasses.

it's not a recursion, because the stack doesn't overflow, fwiw.

**Comment by dkg** [comment]
the problem is in `pgpy/packet/subpackets/signature.py` in `ByteFlag::parse`, where we have a `self.header.length` that is returning 1571993290 -- so it's not an infinite loop, but it's way too much.

**Comment by Commod0re** [comment]
aha I _just_ figured it out, it's a parse order issue (it's trying to parse the length as the wrong style of length because it hasn't actually detected the length format yet). I've got a fix I'll publish as v0.5.1 shortly

**Comment by Commod0re** [comment]
actually, scratch that, it looks like the issue is it got misaligned by that `type 0` subpacket like you were thinking

**Comment by Commod0re** [comment]
AH HERE IT IS

```python
    def parse(self, packet):  # pragma: no cover
        if self.header._typeid == 0:
            self.header.parse(packet)
```

it really is looping infinitely on this

**Comment by Commod0re** [comment]
anyway, fix followed by v0.5.1 release incoming shortly

**Comment by Commod0re** [comment]
this should now be fixed with the advent of PGPy v0.5.1:
```python
In [1]: import pgpy                                                                                                                 

In [2]: pgpy.__version__                                                                                                            
Out[2]: '0.5.1'

In [3]: pgpy.PGPKey.from_blob(b'\x99\x01\r\x04P\xaa+h\x01\x08\x00\xa2\xd3\xb4\xb5\n\xab\xd5Z\xed\xd6\xf3\x1eQ^\xb3\xddJ\xd4\xbbD\xb1
   ...: \x90\xa4\xe1q\xbb@\xf2=\xb9g\xb0B\x8e/\xc4!\xa97\xd28\xa6\xd9\xdc\xd5|\xb1\xe5\x9c\x08\x98\xd5j\xa9\xb1.,4~?B\x8b}\x83Z\xf3R
   ...: \xa0\x96M\x99 \xd6\x9a\x95|\x81\xc8J\x90:\xb2\xa0\xd1\xca\x0fl\xf0\xd1\xe0R\x933\xa7\x83\xe1\n\xd4y\x12\xf3\xd0A@\xb2\xb2\x0
   ...: 5E2\xeaA^\x8bt\xfe7\xc0\xd7a\xc8\xc4\x1c\xd8\xab\xce\xee\x16D\x85\x8f,}\x85d}!]zM\xbc\xab\xefH\x18\x85aQ\xa5Q\xb655"w\xcb\xb
   ...: 3\xde\x7f\xecm\xe9\xc0\x95\x12V\x7fc1n\xc9V\xac\xe5\x97\x8d\x9a\xd8\xa10"\xa1?_\xd8\x98\xb4N\x99\xb8\xcdUkq\x95\xdby\xf7\xe6
   ...: n\x8eV\xe9-5{\xc7\xe1a\x1d\xf2e\xcb_|Q\xf6\x1a\xd7\x96\x95\xdb\xf1\xea\xea\xccl\x9d\xac\x84/\xc4\x1ez\xb4t\x9a\x8c\xaa\x08;U
   ...: \xd3\xddD;\xea\x9d}\x02\x02\rsM\t\x0b\xb7\x00\x11\x01\x00\x01\xb41Standardsoftware <standardsoftware@volkswagen.de>\x89\x01%
   ...: \x04\x13\x01\x02\x00\x19\x05\x02P\xaa+h\t\x10\x00\xf7\xf3aM\x98xv\x02\x00\t\x05\t\x03\xc2g\x00\x00\x00\r\xb7\x07\xff]\xb2\xb
   ...: 6\xca\x1e\xdc\xbf\x1e\x83\xd0Y\xbbgs\xcf\x92\x1dP\x98!\xe4w\xc7\rrK:Ef\x04\xc0\x96V~\xc1\x83\xa3o\x0b\x98\x95<l2RZ\xaa\xc0\r
   ...: \xa7\xa6\xdfTs\xa9\rR-\x05!\x0c\xc4\xd4\xcf\xb0B\xb7\xaeR\xae\xc6\x0cM\xf8\x80\xf2\x80\xf6%Wu\xf2\x1c-\xa3\nO\xc3\x9ba\x0e \
   ...: xf3~,\xf6.d\xb2\xe0v\xb8\tR%\xf4\x01U\xffi\x8a\xad\x90\xd7\xb0O\xd5%[V\xbb6\xa1\x13\xa0{\xf0\xd2\x14X\xa61\xdd\xf0 \xb2j\xcb
   ...: ko\xaf\x9b\x7fp3\x17\x7f\x85\xbd\xc1Y7\xf8e\x08\x84\x81V\xd2T\xa7\x91\xeaw\x8f\x1f\xc5\x9cX\xf3\xcd\x96\x8d\xa6p2fR\xf7[|\x1
   ...: 2\xff\xab>\xe2\xd4\x12j\xf9[y\xcc\xc3\xc0\xe4L\xf3\xc0\x00\x95z\x98\xcdx\xb6\x14\xcar#T\xc7\xb3 x~\x8dD\xcby\xd5\xa7\xe0[\xb
   ...: 7 \x1fi{\xcfo\x86i\xbc\x81\x0b\xaf\x0eE\x89p\xcd\xdd\xb4\xf8\xad\xf2\x02Yb\xf0\x02\xb9\xd2\x92\x02')                        
Out[3]: 
(<PGPKey [PubKeyV4][0x00F7F3614D987876] at 0x7F1451C265C0>,
 OrderedDict([(('00F7F3614D987876', True),
               <PGPKey [PubKeyV4][0x00F7F3614D987876] at 0x7F1451C265C0>)]))

```
